icicle/appUtils/ntt/ntt.cu

```diff
@@ -385,34 +388,68 @@ namespace ntt {
     // but it's a singleton that is supposed to be initialized once per program lifetime
     if (!Domain<S>::twiddles) {
+      bool found_logn = false;
       S omega = primitive_root;
       unsigned omegas_count = S::get_omegas_count();
-      for (int i = 0; i < omegas_count; i++)
+      for (int i = 0; i < omegas_count; i++) {
         omega = S::sqr(omega);
+        if (!found_logn) {
+          ++Domain<S>::max_log_size;
+          found_logn = omega == S::one();
+        }
+      }
+      Domain<S>::max_size = (int)pow(2, Domain<S>::max_log_size);
       if (omega != S::one()) {
-        std::cerr << "Primitive root provided to the InitDomain function is not in the subgroup" << '\n';
-        throw -1;
+        throw IcicleError(
+          IcicleError_t::InvalidArgument, "Primitive root provided to the InitDomain function is not in the subgroup");
       }
 
-      std::vector<S> h_twiddles;
-      h_twiddles.push_back(S::one());
-      int n = 1;
-      do {
-        Domain<S>::coset_index[h_twiddles.at(n - 1)] = n - 1;
-        h_twiddles.push_back(h_twiddles.at(n - 1) * primitive_root);
-      } while (h_twiddles.at(n++) != S::one());
-
-      CHK_IF_RETURN(cudaMallocAsync(&Domain<S>::twiddles, n * sizeof(S), ctx.stream));
-      CHK_IF_RETURN(
-        cudaMemcpyAsync(Domain<S>::twiddles, &h_twiddles.front(), n * sizeof(S), cudaMemcpyHostToDevice, ctx.stream));
-
-      Domain<S>::max_size = n - 1;
+      // allocate and calculate twiddles on GPU
+      // Note: radix-2 INTT needs ONE in last element (in addition to first element), therefore have n+1 elements
+      // Managed allocation allows host to read the elements (logn) without copying all (n) TFs back to host
+      CHK_IF_RETURN(cudaMallocManaged(&Domain<S>::twiddles, (Domain<S>::max_size + 1) * sizeof(S)));
+      CHK_IF_RETURN(generate_external_twiddles_generic(
+        primitive_root, Domain<S>::twiddles, Domain<S>::internal_twiddles, Domain<S>::basic_twiddles,
+        Domain<S>::max_log_size, ctx.stream));
       CHK_IF_RETURN(cudaStreamSynchronize(ctx.stream));
+
+      const bool is_map_only_powers_of_primitive_root = true;
+      if (is_map_only_powers_of_primitive_root) {
+        // populate the coset_index map. Note that only powers of the primitive-root are stored (1, PR, PR^2, PR^3...)
+        Domain<S>::coset_index[S::one()] = 0;
+        for (int i = 0; i < Domain<S>::max_log_size; ++i) {
+          const int index = (int)pow(2, i);
+          Domain<S>::coset_index[Domain<S>::twiddles[index]] = index;
+        }
+      } else {
+        // populate all values
+        for (int i = 0; i < Domain<S>::max_size; ++i) {
+          Domain<S>::coset_index[Domain<S>::twiddles[i]] = i;
+        }
+      }
     }
 
     return CHK_LAST();
   }
 
+  template <typename S>
+  cudaError_t Domain<S>::ReleaseDomain(device_context::DeviceContext& ctx)
+  {
+    CHK_INIT_IF_RETURN();
+
+    max_size = 0;
+    max_log_size = 0;
+    cudaFreeAsync(twiddles, ctx.stream);
```

Comment:
What if we call ReleaseDomain on the non-allocated domain? Maybe worth adding a check

Reply:
I initially did but I checked and it behaves like free(nullptr) which is valid. So doing cudaFreeAsync(nullptr) is not a problem.